backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/Autoboxing.kt

```diff
@@ -309,6 +309,24 @@ private class InlineClassTransformer(private val context: Context) : IrBuildingT
         return declaration
     }
 
+    override fun visitCall(expression: IrCall): IrExpression {
+        expression.transformChildrenVoid(this)
+        if (context.shouldOptimize()) {
+            val property = expression.symbol.owner.correspondingPropertySymbol?.owner ?: return expression
+
+            property.parent.let {
+                if (it is IrClass && it.isInline && property.backingField != null) {
+                    expression.dispatchReceiver?.let {
+                        it.type = expression.type
```

Comment:
This might be harmful, because resulting expression would have type that doesn't match its "inferred" type.

Reply:
But what you suggest, make a copy?

Reply:
I suggest returning it as is.

Reply:
Then we need to return to the first version and make separate lowering to run it after devirtualization. Should I revert last commits?

Reply:
> Then we need to return to the first version and make separate lowering to run it after devirtualization

Why do you find this necessary? What exactly goes wrong if we don't patch the type here?

Reply:
https://github.com/JetBrains/kotlin-native/blob/b4d5f092a138b9aed2f459e44774c5c0ef18cca7/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/optimizations/Devirtualization.kt#L1463

In case 
```
inline class A(val data: String)

...

if (A("a").data == "b") ..
```

we can't find devirtualized function equals, because if don't change type we are trying to find `equals(A, String)` and get wrong receiver type. DFG for this call has wrong receiver type. If stay type we can make this change only after building DFG phase.

Reply:
I see.
Well, in this case using `reinterpret` intrinsic should help. See the example here: https://github.com/JetBrains/kotlin-native/blob/e4d75ccf41cba6116f85269f478cd951c9c4acde/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/Autoboxing.kt#L259

Also I guess we might have more problems like this, e.g. due to https://github.com/JetBrains/kotlin-native/blob/e4d75ccf41cba6116f85269f478cd951c9c4acde/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/Autoboxing.kt#L440

Reply:
Ok, I was trying not to use extra calls.